alchemist/alchemist-incarnation-biochemistry/src/main/java/it/unibo/alchemist/model/implementations/actions/ChemiotacticPolarization.java

```diff
@@ -0,0 +1,117 @@
+        }
+        this.env = environment;
+        this.biomol = new Biomolecule(biomol);
+        if (ascendGrad.equals("up")) {
+            this.ascend = true;
+        } else if (ascendGrad.equals("down")) {
+            this.ascend = false;
+        } else {
+            throw new IllegalArgumentException("Possible imput string are only up or down");
+        }
+    }
+
+    @Override
+    public ChemiotacticPolarization cloneOnNewNode(final Node<Double> n, final Reaction<Double> r) {
+        return new ChemiotacticPolarization(env, n, biomol.toString(), ascend ? "up" : "down");
+    }
+
+    @Override
+    public void execute() {
+        final List<Node<Double>> l = env.getNeighborhood(getNode()).getNeighbors().stream()
+                .parallel()
+                .filter(n -> n instanceof EnvironmentNode && n.contains(biomol))
+                .collect(Collectors.toList());
+        if (l.isEmpty()) {
+            ((CellNode) getNode()).addPolarization(new Continuous2DEuclidean(0, 0));
+        } else {
+            final boolean isNodeOnMaxConc = env.getPosition(l.stream()
+                    .max((n1, n2) -> Double.compare(n1.getConcentration(biomol), n2.getConcentration(biomol)))
+                    .get()).equals(env.getPosition(getNode()));
+            if (isNodeOnMaxConc) {
+                ((CellNode) getNode()).addPolarization(new Continuous2DEuclidean(0, 0));
+            } else {
+                Position newPolVer = weightedAverageVectors(l);
+                final double newPolVerModule = FastMath.sqrt(FastMath.pow(
+                        newPolVer.getCoordinate(0), 2) + FastMath.pow(newPolVer.getCoordinate(1), 2)
+                        );
+                if (newPolVerModule == 0) {
+                    ((CellNode) getNode()).addPolarization(newPolVer);
+                } else {
+                    newPolVer = new Continuous2DEuclidean(newPolVer.getCoordinate(0) / newPolVerModule, newPolVer.getCoordinate(1) / newPolVerModule);
+                    if (ascend) {
+                        ((CellNode) getNode()).addPolarization(newPolVer);
+                    } else {
+                        ((CellNode) getNode()).addPolarization(new Continuous2DEuclidean(
+                                -newPolVer.getCoordinate(0), 
+                                -newPolVer.getCoordinate(1))
+                                );
+                    }
+                }
+            }
+        }
+    }
+
+    private Position weightedAverageVectors(final List<Node<Double>> list) {
+        Position res = new Continuous2DEuclidean(0, 0);
+        for (final Node<Double> n : list) {
+            Position vecTemp = new Continuous2DEuclidean(
+                    env.getPosition(n).getCoordinate(0) - env.getPosition(getNode()).getCoordinate(0),
+                    env.getPosition(n).getCoordinate(1) - env.getPosition(getNode()).getCoordinate(1));
```

Comment:
Calling getPosition() comes with a cost. Call it **once**, put the result in a variable, then use the variable!